Variation details: Add image-editing functionality to variations

0nko opened the pull request:
Fixes #2765. This PR adds the image-editing functionality to variations, which have a limit of one image. 

![image](https://user-images.githubusercontent.com/1522856/92575323-77f15480-f288-11ea-8c43-b4523189ff4b.png)

Note: Currently, the API doesn't support image removal, so the Trash icon has been removed from the image detail.

**To test:**

1. Go to Products -> Tap on a variable product -> Tap on Variations -> Select a variation without images
2. Tap on the Add image button and notice the image source dialog is displayed
3. Select an image from local storage and notice only one image at a time can be selected
4. Notice a thumbnail with a progress icon is displayed
5. Tap on Done button
6. Notice the thumbnail is also displayed in the variation detail screen
7. After the upload is finished, notice that the progress bar disappears
8. Notice that the Add image button is not visible anymore
9. Tap on the image
10. Notice the image is displayed full-screen and that there's no Remove button
11. Tap on the back button
12. Notice there is a Replace photo button at the top
13. Tap on it and use WP gallery as a source
14. Notice only one image at a time can be selected and after selection, the image is replaced immediately
15. Tap on Replace photo button again
16. Select the camera source
17. After taking a picture, notice the image is uploaded
18. Tap on Done
19. Tap on Update
20. Notice the variation is saved and the Update button is gone

0nko committed:
Add image gallery support to variation details

0nko committed:
Merge branch 'refactor/decouple-image-gallery' into issue/2765-variation-images

0nko committed:
Add new string resources

0nko committed:
Remove unused navigation graph element

0nko committed:
Add a way to disable image multi-selection in WP media picker

0nko committed:
Disable multi-selection of images in local chooser for variations

0nko committed:
Show the warning and update button title for variations

0nko committed:
Replace the old result navigation in WP image picker and return the imges to the image gallery

0nko committed:
Only show a single image in the gallery for variations

0nko committed:
Hide the add image icon in varitions if there already is one

0nko committed:
Merge branch 'develop' into issue/2765-variation-images

# Conflicts:
#	WooCommerce/src/main/kotlin/com/woocommerce/android/ui/products/ProductImagesFragment.kt

0nko committed:
Remove unused code

0nko committed:
Update the state of uploading and uploaded images in variation details

0nko committed:
Fix ktlint and checkstyle issues

0nko committed:
Fix the discard dialog message/title argument discrepancy

0nko committed:
Disable the update button if images are being uploaded

0nko committed:
Fix ktlint issues

nbradbury commented:
@0nko When I replace a variation's photo with an image from the WP media library and return to the variation detail, the Update button remains grayed out. However, if I try to leave detail, I get the Discard dialog.

![Screenshot_1599762803](https://user-images.githubusercontent.com/3903757/92783259-9c762b00-f373-11ea-82d7-237d538a32f0.png)

nbradbury commented:
@0nko @Garance91540 When viewing a variation that has no image, I'm wondering if the text should read "Add a variation image" instead of "Add a product image."

![Screenshot_1599763843](https://user-images.githubusercontent.com/3903757/92785252-3f7b7480-f375-11ea-8358-8f70aaf4d712.png)

nbradbury commented:
@0nko Perhaps we should update the WP media picker to work like the device picker and return to the previous screen as soon as an image is tapped? It seems odd to be limited to a single image, but still show "1 selected" and require the user to tap Done.

![Screenshot_1599764293](https://user-images.githubusercontent.com/3903757/92786606-638b8580-f376-11ea-9007-b398f08fc910.png)

nbradbury commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/products/ProductImagesFragment.kt:
```diff
@@ -162,15 +178,15 @@ class ProductImagesFragment : BaseProductEditorFragment(R.layout.fragment_produc
             it.type = "image/*"
-            it.putExtra(Intent.EXTRA_ALLOW_MULTIPLE, true)
             it.addCategory(Intent.CATEGORY_OPENABLE)
+            it.putExtra(Intent.EXTRA_ALLOW_MULTIPLE, viewModel.isLimitedToSingleImage)
```
This appears to be inverted. When called from "normal" product detail, `isLimitedToSingleImage` is `true`. When called from variation detail, it's `false`.

0nko replied:
Ah, good catch! I renamed the variable from `isProduct`, so the boolean got inverted.

nbradbury commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/products/ProductImagesViewModel.kt:
```diff
@@ -41,24 +41,26 @@ class ProductImagesViewModel @AssistedInject constructor(
             uploadingImageUris = ProductImagesService.getUploadingImageUris(navArgs.remoteId),
-            isImageDeletingAllowed = navArgs.requestCode == RequestCodes.PRODUCT_DETAIL_IMAGES,
-            images = navArgs.images.toList()
+            isImageDeletingAllowed = isLimitedToSingleImage,
```
This is also inverted.

nbradbury commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/products/ProductImagesViewModel.kt:
```diff
@@ -120,6 +122,14 @@ class ProductImagesViewModel @AssistedInject constructor(
     }
 
+    fun onImagesAdded(newImages: List<Image>) {
+        viewState = if (isLimitedToSingleImage) {
```
Also inverted. If we're limited to a single image, we don't want to add the new image to the existing images. However, because `isLimitedToSingleImage` is incorrectly `true` for normal product detail, it's working as expected. 

I see similar problems elsewhere, too, but I don't think I need to point them out. Maybe simply rename `isLimitedToSingleImage` to `isMultiSelectionAllowed `?

Garance91540 commented:
> When viewing a variation that has no image, I'm wondering if the text should read "Add a variation image" instead of "Add a product image."

Sounds right to me @nbradbury

0nko committed:
Change the title of the Add image in variation details

(cherry picked from commit 81a2adf006292e479ef5fb9a57753abe690263f7)

0nko committed:
Merge branch 'develop' into issue/2765-variation-images

0nko committed:
Rename the multi-image selection flag to refelct the actual state

0nko committed:
Return from the WP image picker immediately if only single photo allowed

0nko committed:
Display the image gallery instead of an image detail after the image is clicked in product/variation detail

0nko committed:
Enable the Update button by default (for WP image picker)

0nko committed:
Optimize imports

0nko commented:
Thanks for the review, @nbradbury! I've addressed all issues you've mentioned. I also changed the image gallery click behavior to open the gallery screen instead of the image detail.

Ready for another round.

nbradbury commented:
@0nko Changes look good, but I think we should remove the circle overlay and "Done" button when limited to single select.

![Screenshot_1599825120](https://user-images.githubusercontent.com/3903757/92922510-f0305500-f403-11ea-8cf3-7a28258e0a08.png)

0nko committed:
Remove the selection count if multi-selection disabled

0nko committed:
Hide the Update button when multi-selection disabled

0nko commented:
Yeah, good call. It's fixed now.

nbradbury reviewed (approved):
Looks great! :shipit:
